servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpRequestTracker.java

```diff
@@ -0,0 +1,108 @@
+        return filter.create(connection);
+    }
+
+    private static class Observer implements HttpLifecycleObserver {
+
+        private final Function<HttpResponseMetaData, ErrorClass> peerResponseErrorClassifier;
+        private final Function<Throwable, ErrorClass> errorClassFunction;
+        private final RequestTracker tracker;
+
+        Observer(final Function<HttpResponseMetaData, ErrorClass> peerResponseErrorClassifier,
+                 final Function<Throwable, ErrorClass> errorClassFunction, final RequestTracker tracker) {
+            this.peerResponseErrorClassifier = peerResponseErrorClassifier;
+            this.errorClassFunction = errorClassFunction;
+            this.tracker = tracker;
+        }
+
+        @Override
+        public HttpExchangeObserver onNewExchange() {
+            return new RequestTrackerExchangeObserver();
+        }
+
+        private final class RequestTrackerExchangeObserver implements HttpLifecycleObserver.HttpExchangeObserver {
+
+            // TODO: cleanup.
+            private final AtomicLong startTime = new AtomicLong(Long.MIN_VALUE);
+
+            @Override
+            public void onConnectionSelected(ConnectionInfo info) {
+                // noop
+            }
+
+            @Override
+            public HttpLifecycleObserver.HttpRequestObserver onRequest(HttpRequestMetaData requestMetaData) {
+                startTime.set(tracker.beforeRequestStart());
+                return NoopHttpLifecycleObserver.NoopHttpRequestObserver.INSTANCE;
+            }
+
+            @Override
+            public HttpLifecycleObserver.HttpResponseObserver onResponse(HttpResponseMetaData responseMetaData) {
+                ErrorClass error = peerResponseErrorClassifier.apply(responseMetaData);
+                if (error != null) {
+                    final long startTime = finish();
+                    if (checkOnce(startTime)) {
+                        tracker.onRequestError(startTime, error);
+                    }
+                }
+                return NoopHttpLifecycleObserver.NoopHttpResponseObserver.INSTANCE;
+            }
+
+            @Override
+            public void onResponseError(Throwable cause) {
+                final long startTime = finish();
+                if (checkOnce(startTime)) {
+                    tracker.onRequestError(startTime, errorClassFunction.apply(cause));
+                }
+            }
+
+            @Override
+            public void onResponseCancel() {
+                final long startTime = finish();
+                if (checkOnce(startTime)) {
+                    tracker.onRequestError(startTime, ErrorClass.CANCELLED);
+                }
+            }
+
+            @Override
+            public void onExchangeFinally() {
+                final long startTime = finish();
+                if (checkOnce(startTime)) {
+                    tracker.onRequestSuccess(startTime);
```

Comment:
`onExchangeFinally()` doesn't mean "success". It only means that now it's time to create a summary based on previously invoked callbacks like `onResponseComplete/onResponseError/onResponseCancel`. Reason is because there is a chance the response will be received before the request completes. For example, server can response with `202 Accepted` and client keeps sending large request.

Reply:
In that case, consider using `onResponseComplete` instead of `onExchangeFinally`, there is no need to defer successful completion if you don't defer failure/cancel